### DISCUSSION TITLE
How to customize and add LLM models

### DISCUSSION THREAD

**Discussion by smileyboy2019** [langchain4j/langchain4j#2174] [Q&A]
How to customize and add LLM models? Now there is a new LLM, how to configure it through customization. The main parameters are API (Domain) URL and API key. If so, can we customize the model to be added

**Comment by smileyboy2019** [comment]
How to add Kimi large model  https://api.moonshot.cn/v1

**Comment by InAnYan** [comment]
@langchain4j some strange stuff happening in this thread

**Comment by InAnYan** [comment]
@smileyboy2019 if you want to support a new LLM in your app, there are two ways based on your LLM:
1) If it's an LLM that is from the same company, that you used before, then you can just supply a new model name in the, for example `OpenAiChatModel.builder().modelName("...")`. This parameter is typically a `String`, so in case current version of `langchain4j` doesn't support it (in the enum), then you can always pass a string.

2) If it's an LLM of some other company, then look through [langchain4j integrations](https://docs.langchain4j.dev/category/integrations). Find the company there. If it is there, then follow the examples.

3) If it's an LLM of some other company, that langchain4j currently doesn't support, then you are a bit out of luck. You either need to implement it manually (you've sent a link, so  they have an API), or: if they have OpenAI compatible API, then you can reuse `OpenAiChatModel` but with custom API base URL and model name

(I'm leaving this comment just for history, or if someone would collect FAQ, and my answer would be useful)

**Reply by evelynn996** [reply]
thanks

**Comment by InAnYan** [comment]
@smileyboy2019 please don't click on the links provided by seeronline and smileyboy2019, they are not relevant and may contain malware

**Comment by smileyboy2019** [comment]
If the model only allows API URLs and API keys to be provided, can it be called through openAPI
![image](https://github.com/user-attachments/assets/6f02e7be-e01b-4149-8a69-ca76f15d9808)

**Comment by InAnYan** [comment]
> If the model only allows API URLs and API keys to be provided, can it be called through openAPI ![image](https://private-user-images.githubusercontent.com/59221294/361393852-6f02e7be-e01b-4149-8a69-ca76f15d9808.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ2NzAwNDEsIm5iZiI6MTcyNDY2OTc0MSwicGF0aCI6Ii81OTIyMTI5NC8zNjEzOTM4NTItNmYwMmU3YmUtZTAxYi00MTQ5LThhNjktY2E3NmYxNWQ5ODA4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODI2VDEwNTU0MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU2ZjBkZDM4ZWJmYjc2MmM1ZmI3Y2RlOTQwNzMwNjk4ODlhNzNhYTdkNmY2NzY0NmI4N2QyZGJjYTg5ZWM4NmYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Vcy9CEVjZ6X9mvJNjubWKnrxaFv0ofTzv15W8QqaqOk)

Of course, you can. But be sure that the API is OpenAI API compatible
